### ISSUE TITLE
Make the fleet listing sortable by ship type

### ISSUE THREAD

**Issue by muglyObserver** [endless-sky/endless-sky#3201]
When your list of ships is several pages long on the Player Info screen, it can be difficult to keep things organized.  It would be nice if you could click on the headings (i.e. "ship", "model", "system", etc) to sort the list by those values.  I would be Okay if the list reverts to the "natural" order if you exit and come back.  That might actually be better than permanently sorting the list.

**Comment by Amazinite** [comment]
FYI: You can edit your issues instead of needing to make a new one because you sent it too soon.

**Comment by muglyObserver** [comment]
Thank you.  I figured that out right after it closed it.  :-)

**Comment by endless-sky** [comment]
Note that if the list were sorted in any way, it wouldn't be possible to change your flagship by dragging a ship to the top. (In fact, it wouldn't be possible to drag ships around at all.) I'm worried that the confusion that would add for the player would cancel out any benefit.

**Comment by hellminister** [comment]
we could have also a "custom" sort that keeps the order of the as set by the player that could also be used for the flagship

**Comment by tehhowch** [comment]
Care to elaborate more clearly, @hellminister ? 
This phrase:
> keeps the order of the as set by the player

does not make any sense to me.

**Comment by hellminister** [comment]
just a setting to keep the ordering in the same way as it is at the moment in the game (without the sorting)

(the missing word is ships)

in other words, have an unsorted setting that lets us place the ships as we like, like at the moment

the only change for this needed in the save file would be a line that gives the sorting type currently used, and the actual sorting is done programatically, (so no need to save the new order based on the sorting choice)

**Comment by tehhowch** [comment]
What MZ had said above is that such a method (programmatic sort) inherently means the player can no longer drag to reorder if they sort the list. The confusion caused by such a change in behavior (I can't change my flagship!!) needs to be mitigated somehow, and the only benefit is finding a ship faster. Once you've found a ship in this sorted mode, you can only park /disown it, or open its detail view (which would also need modification to use a different list order when clicking next and previous).

**Comment by jafdy** [comment]
Always putting the flag ship at the top would let you always find it, giving it an extra box for it's self would also help.
It's won't let the player sort by dragging though, unless it was only a one-time sort that only puts ships in the requested order.

**Comment by KiLEdEnNis** [comment]
what if the I menu got an extra button in Ship Info: "Set Flagship", and a * or something before its name on the Player Info tab?
you no longer have to have your flagship as the 1st non-automaton/parked/carried ship in the list, so you can freely sort it

**Comment by postscripter8** [comment]
`Set Flagship` button could do the trick. Keeping the flagship first can be preserved by a simple hack in the comparison function(s) implementation(s), like e.g.:

```
int compareShips(Ship * s1, Ship * s2) {
  int flg = s1->isFlagship() - s2->isFlagship();
  if (flg) return flg;
  // other criteria follows
}
```

**Comment by asl97** [comment]
The title says sortable instead of sorted, as well as a number of people has suggested a button for a one time sort and somehow we reach the conclusion of a set-flagship button :facepalm: .

While "sorting" usually implies sorting from A to Z,  it doesn't and probably shouldn't be done that way,  it's better to sort it by the order ships appear in as suggested by @hellminister.

This way, the flagship stays on top and your main fleet ships which should be after the flagship also remain near the top.

<details>
<summary>Sorting algorithm example with data</summary>

```python
def sorting_ship(ships):
    types = {}
    types_order = []
    for ship in ships:
        type = ship['type']
        if type not in types:
            types[type] = []
            types_order.append(type)
        types[type].append(ship)
    ordered = []
    for type in types_order:
        ordered.extend(types[type])
    return ordered

example_data=[{'name': '"Bruja the Greasy"', 'parked': False, 'type': 'Firebird'},
 {'name': '"Black Doom"', 'parked': False, 'type': 'Hawk'},
 {'name': '"Simmering Judas"', 'parked': False, 'type': 'Sparrow'},
 {'name': '"Draco\'s Bolt"', 'parked': True, 'type': 'Sparrow'},
 {'name': '"Flying Darkness"', 'parked': False, 'type': '"Modified Argosy"'},
 {'name': 'Ravager', 'parked': True, 'type': 'Fury'},
 {'name': '"Profane Blade"', 'parked': True, 'type': 'Fury'},
 {'name': '"Maniac of Torture"', 'parked': True, 'type': 'Hawk'},
 {'name': '"Abaddon the Wicked"', 'parked': True, 'type': 'Sparrow'},
 {'name': '"Wench of Hatred"', 'parked': False, 'type': 'Sparrow'},
 {'name': 'Deathwrought', 'parked': True, 'type': 'Sparrow'},
 {'name': 'Sword', 'parked': True, 'type': 'Sparrow'},
 {'name': '"End of Children"', 'parked': True, 'type': 'Falcon'},
 {'name': '"No Mother"', 'parked': True, 'type': 'Wasp'},
 {'name': '"Vlad the Impaler"', 'parked': True, 'type': 'Firebird'},
 {'name': '"Psychotic Angels"', 'parked': True, 'type': 'Sparrow'}]

import pprint
pprint.pprint(sorting_ship(example_data))

```
</details>

Data extracted from the bottom of my ship list and somewhat randomly marking a few as unparked.
Status of park doesn't and probably shouldn't affect the sort.

|   Before    |   Parked   |     Type     |     After     |   Parked   |     Type     | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
|"Bruja the Greasy"|False|Firebird|"Bruja the Greasy"|False|Firebird|
|"Black Doom"|False|Hawk|"Vlad the Impaler"|True|Firebird|
|"Simmering Judas"|False|Sparrow|"Black Doom"|False|Hawk|
|"Draco's Bolt"|True|Sparrow|"Maniac of Torture"|True|Hawk|
|"Flying Darkness"|False|"Modified Argosy"|"Simmering Judas"|False|Sparrow|
|Ravager|True|Fury|"Draco's Bolt"|True|Sparrow|
|"Profane Blade"|True|Fury|"Abaddon the Wicked"|True|Sparrow|
|"Maniac of Torture"|True|Hawk|"Wench of Hatred"|False|Sparrow|
|"Abaddon the Wicked"|True|Sparrow|Deathwrought|True|Sparrow|
|"Wench of Hatred"|False|Sparrow|Sword|True|Sparrow|
|Deathwrought|True|Sparrow|"Psychotic Angels"|True|Sparrow|
|Sword|True|Sparrow|"Flying Darkness"|False|"Modified Argosy"|
|"End of Children"|True|Falcon|Ravager|True|Fury|
|"No Mother"|True|Wasp|"Profane Blade"|True|Fury|
|"Vlad the Impaler"|True|Firebird|"End of Children"|True|Falcon|
|"Psychotic Angels"|True|Sparrow|"No Mother"|True|Wasp|

PS, I honestly don't really need a sorting button if I could just select all ship of specific types

**Comment by petervdmeer** [comment]
> Note that if the list were sorted in any way, it wouldn't be possible to change your flagship by dragging a ship to the top. (In fact, it wouldn't be possible to drag ships around at all.)

This assumes that the sorting-action itself is persistent. I would expect that the sort-action is executed only once each time the sort header-button is pressed. The (selected) ships will be stored in the sorted order in the ships-list, but the player can still do all the re-ordering he/she wants after sorting. Ships bought or captured will still always be added to the end of the list.

The reason that I do want sort to update the ship-list itself (instead of just temporary displaying a different order) is that I might want to sell all the ships with most crew-requirements, or that i might want to sell all my Sparrows. If sorting only works on the player-info screen, then I cannot use the results of sorting elsewhere.

**Comment by Amazinite** [comment]
Resolved by #5574.
